File: Tutorials/WinForms/DevExpress.MVVM/CS/ViewModels/CustomerListViewModel.cs
using System.ComponentModel;
using DevExpress.Data.Filtering;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Xpo;
using WinFormsApplication.Services;
using XpoTutorial;

namespace WinFormsApplication.ViewModels {
    public class CustomerListViewModel {
        public virtual IListSource Customers {
            get;
            protected set;
        }
        protected IInstantFeedbackService InstantFeedbackService {
            get { return this.GetService<IInstantFeedbackService>(); }
        }
        void ReloadCore() {
            Customers = Customers ?? new XPInstantFeedbackView(typeof(Customer), new ServerViewProperty[] {
                    new ServerViewProperty("Oid","Oid"),
                    new ServerViewProperty("ContactName", SortDirection.Ascending, new OperandProperty("ContactName"))
                }, null);
            ((XPInstantFeedbackView)Customers).Refresh();
        }
        public void Reload() {
            if(InstantFeedbackService == null)
                ReloadCore();
            else {
                object focusedObjectKey = InstantFeedbackService.GetFocusedRowKey();
                ReloadCore();
                InstantFeedbackService.SetFocusedRowByKey(focusedObjectKey);
            }
        }
        public async void Delete() {
            using(UnitOfWork uow = new UnitOfWork()) {
                object key = InstantFeedbackService.GetFocusedRowKey();
                Customer customer = uow.GetObjectByKey<Customer>(key);
                uow.Delete(customer);
                await uow.CommitChangesAsync();
            }
            Reload();
        }
        public void Edit() {
            CreateDocument((int)InstantFeedbackService.GetFocusedRowKey());
        }
        public void Add() {
            CreateDocument(-1);
        }
        //
        protected IDocumentManagerService DocumentManagerService {
            get { return this.GetService<IDocumentManagerService>(); }
        }
        protected void CreateDocument(int customerId) {
            var id = new ViewID(typeof(Customer), customerId);
            IDocument document = DocumentManagerService.FindDocumentByIdOrCreate(id, svc => {
                var doc = svc.CreateDocument("EditCustomerView", customerId, this);
                doc.Id = id;
                doc.Title = string.Format("Edit Customer {0}", customerId);
                return doc;
            });
            document.Show();
        }
    }
}

File: Tutorials/WinForms/DevExpress.MVVM/VB/ViewModels/CustomerListViewModel.vb
Imports System.ComponentModel
Imports DevExpress.Data.Filtering
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpo
Imports WinFormsApplication.Services
Imports XpoTutorial

Namespace WinFormsApplication.ViewModels
	Public Class CustomerListViewModel
		Public Sub New()
			Customers = New XPInstantFeedbackView(GetType(Customer), New ServerViewProperty() {
				New ServerViewProperty("Oid", "Oid"),
				New ServerViewProperty("ContactName", SortDirection.Ascending, New OperandProperty("ContactName"))
			}, Nothing)
		End Sub
		Private privateCustomers As IListSource
		Public Overridable Property Customers() As IListSource
			Get
				Return privateCustomers
			End Get
			Protected Set(ByVal value As IListSource)
				privateCustomers = value
			End Set
		End Property
		Protected ReadOnly Property InstantFeedbackService() As IInstantFeedbackService
			Get
				Return Me.GetService(Of IInstantFeedbackService)()
			End Get
		End Property
		Private Sub ReloadCore()
			DirectCast(Customers, XPInstantFeedbackView).Refresh()
		End Sub
		Public Sub Reload(Optional ByVal recordToFocus? As Integer = Nothing)
			If InstantFeedbackService Is Nothing Then
				ReloadCore()
			Else
				Dim focusedObjectKey As Object = If(recordToFocus.HasValue, recordToFocus.Value, InstantFeedbackService.GetFocusedRowKey())
				ReloadCore()
				InstantFeedbackService.SetFocusedRowByKey(focusedObjectKey)
			End If
		End Sub
		Public Async Sub Delete()
			Using uow As New UnitOfWork()
				Dim key As Object = InstantFeedbackService.GetFocusedRowKey()
				Dim customer As Customer = uow.GetObjectByKey(Of Customer)(key)
				uow.Delete(customer)
				Await uow.CommitChangesAsync()
			End Using
			Reload()
		End Sub
		Public Sub Edit()
			CreateDocument(DirectCast(InstantFeedbackService.GetFocusedRowKey(), Integer))
		End Sub
		Public Sub Add()
			CreateDocument(-1)
		End Sub
		'
		Protected ReadOnly Property DocumentManagerService() As IDocumentManagerService
			Get
				Return Me.GetService(Of IDocumentManagerService)()
			End Get
		End Property
		Protected Sub CreateDocument(ByVal customerId As Integer)
			Dim id = New ViewID(GetType(Customer), customerId)
			Dim document As IDocument = DocumentManagerService.FindDocumentByIdOrCreate(id, Function(svc)
				Dim doc = svc.CreateDocument("EditCustomerView", customerId, Me)
				doc.Id = id
				doc.Title = String.Format("Edit Customer {0}", customerId)
				Return doc
			End Function)
			document.Show()
		End Sub
	End Class
End Namespace
